pkg/network/ebpf/c/protocols/http/http.h

```diff
@@ -146,9 +133,17 @@ static __always_inline int http_process(http_transaction_t *http_stack, skb_info
         http->response_last_seen = bpf_ktime_get_ns();
     }
 
-    if (http_closed(http, skb_info, http_stack->owned_by_src_port)) {
+    if (http_closed(skb_info)) {
         http_batch_enqueue(http);
-        bpf_map_delete_elem(&http_in_flight, &http_stack->tup);
+        long ret = bpf_map_delete_elem(&http_in_flight, &http_stack->tup);
+        if (ret < 0) {
+            // unlikely: this means that we're dealing with a localhost request
+            // whose FIN packets were processed concurrently; to avoid double
+            // flushing it, we use the return value of `bpf_map_delete_elem` as
+            // a synchronization lock and remove the element from the batch if
+            // the deletion fails;
+            http_batch_pop();
+        }
```

Comment:
if i understand correctly, then the following scenario can happen as well:
we are processing another request, and we are enqueueing it just before we are calling `pop`, which lead us to remove the wrong event

Reply:
all batch operations target a specific CPU. the scenario you're describing would require a socket filter program to be preempted AND followed by the the execution of another socket filter program in the same CPU which is not possible AFAIK.

also, the execution of this code path is truly unlikely. usually what happens in the context of localhost traffic is something like

1. FIN packet from server to client
    `http_process()` is executed and results in flushing the httpTX to userpace and deleting the entry from `http_in_flight`

which is _followed_ by

2. FIN packet from client to server
http_process() fails early because there isn't even a httpTX matching the tuple

`http_batch_pop` is really a safeguard only meant to prevent a double flush when 1 and 2 happen concurrently

Reply:
Regarding preemption
that might be possible https://lore.kernel.org/bpf/CAMy7=ZWPc279vnKK6L1fssp5h7cb6cqS9_EuMNbfVBg_ixmTrQ@mail.gmail.com/T/#mc845d9ceb334335f02a6c110320e2e3891f5a5bb

Reply:
But I get the answer